app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt

```diff
@@ -568,52 +573,44 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
         val sessionManager = requireComponents.core.sessionManager
 
         // Update the UI with the tab removed, but don't remove it from storage yet
-        getManagedEmitter<SessionControlChange>().onNext(
-
-            SessionControlChange.TabsChange(
-                sessionManager.sessions
-                    .filter { (activity as HomeActivity).browsingModeManager.isPrivate == it.private }
-                    .filter { it.id != sessionId }
-                    .map {
-                        val selected = it == sessionManager.selectedSession
-                        it.toTab(context!!, selected)
-                    }
-            )
-        )
+        emitSessionChanges(sessionId)
 
         val deleteOperation: (suspend () -> Unit) = {
             sessionManager.findSessionById(sessionId)
                 ?.let { session ->
+                    pendingSessionDeletion = null
                     sessionManager.remove(session)
                 }
         }
 
-        deleteSessionJob = deleteOperation
+        pendingSessionDeletion = PendingSessionDeletion(deleteOperation, sessionId)
 
         allowUndo(
-            view!!, getString(R.string.snackbar_tab_deleted),
+            view!!,
+            getString(R.string.snackbar_tab_deleted),
             getString(R.string.snackbar_deleted_undo), {
-                deleteSessionJob = null
+                pendingSessionDeletion = null
                 emitSessionChanges()
             },
             operation = deleteOperation
         )
     }
 
-    private fun emitSessionChanges() {
+    private fun emitSessionChanges(pendingDeletionSessionId: String? = null) {
         val sessionManager = context?.components?.core?.sessionManager ?: return
```

Comment:
Two things: 

1. Make `sessionManager` an instance variable passed in by constructor or `lateinit`ed in `onViewCreated`, or a property computed each time.
2. Add a method to `sessionManager` called `filteredSessions(private: Boolean, (Session) -> Boolean): Collection<Session>`
3. Add an extension method `fun List<Session>.toTabs(selectedId: String?): List<Tab>`.
3. If `getListOfTabs` hasn't disappeared yet, then pass in the closure to `filteredSessions`.